plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java

```diff
@@ -944,6 +945,336 @@ public void testSettingChangeDataFeedEnabledProperty()
+
+        // Select multiple fields from one row field
+        assertQuery("SELECT nested1.child3, id, nested1.child2 FROM " + tableName, "VALUES (100, 1, 'a'), (200, 2, 'b'), (400, 4, NULL), (NULL, 5, NULL)");
+        assertQuery(
+                "SELECT nested2.child2, nested2.child3, id FROM " + tableName,
+                "VALUES (true, DATE '2023-04-19' , 1), (false, DATE '1990-04-20', 2), (NULL, NULL, 4), (true, NULL, 5)");
+
+        // Select multiple fields from multiple row fields
+        assertQuery(
+                "SELECT id, nested2.child1, nested1.child3, nested2.child2, nested1.child1 FROM " + tableName,
+                "VALUES (1, 10.10, 100, true, 10), (2, 20.20, 200, false, 20), (4, NULL, 400, NULL, 40), (5, NULL, NULL, true, NULL)");
+
+        // Select only nested fields
+        assertQuery("SELECT nested2.child2, nested1.child3 FROM " + tableName, "VALUES (true, 100), (false, 200), (NULL, 400), (true, NULL)");
+
+        assertUpdate("DROP TABLE " + tableName);
+    }
+
+    @Test
+    public void testProjectionPushdownAfterMerge()
+    {
+        String targetTable = "test_projection_pushdown_after_merge_target_" + randomNameSuffix();
+        String sourceTable = "test_projection_pushdown_after_merge_source_" + randomNameSuffix();
+
+        assertUpdate("CREATE TABLE " + targetTable + " (customer VARCHAR, purchases INT, address ROW(city VARCHAR))" +
+                " WITH (partitioned_by = ARRAY['purchases'])");
+        assertUpdate("INSERT INTO " + targetTable + " (customer, purchases, address)" +
+                " VALUES ('Aaron', 5, ROW('Antioch')), ('Bill', 7, ROW('Buena')), ('Carol', 3, ROW('Cambridge')), ('Dave', 11, ROW('Devon'))", 4);
+
+        assertUpdate("CREATE TABLE " + sourceTable + " (customer VARCHAR, purchases INT, address ROW(city VARCHAR))");
+        assertUpdate("INSERT INTO " + sourceTable + " (customer, purchases, address)" +
+                " VALUES ('Aaron', 6, ROW('Arches')), ('Ed', 7, ROW('Etherville')), ('Carol', 9, ROW('Centreville')), ('Dave', 11, ROW('Darbyshire'))", 4);
+
+        @Language("SQL") String sql = "MERGE INTO " + targetTable + " t USING " + sourceTable + " s ON (t.customer = s.customer)" +
+                " WHEN MATCHED AND s.address.city = 'Centreville' THEN DELETE" +
+                " WHEN MATCHED THEN UPDATE SET purchases = s.purchases + t.purchases, address = s.address" +
+                " WHEN NOT MATCHED THEN INSERT (customer, purchases, address) VALUES (s.customer, s.purchases, s.address)";
+
+        assertUpdate(sql, 4);
+
+        assertQuery(
+                "SELECT customer, purchases, address.city FROM " + targetTable,
+                "VALUES ('Aaron', 11, 'Arches'), ('Ed', 7, 'Etherville'), ('Bill', 7, 'Buena'), ('Dave', 22, 'Darbyshire')");
+
+        assertUpdate("DROP TABLE " + sourceTable);
+        assertUpdate("DROP TABLE " + targetTable);
+    }
+
+    @Test
+    public void testProjectionPushdownColumnReorderInSchemaAndDataFile()
```

Comment:
I will move this to the product test Since Trino does not allow creating the table with column mapping mode.

Reply:
Current status:

#### When the data is inserted using `Delta`:

- With `none`: Pass
- With `id`:

  - on Delta:
    ```
      Caused by: java.lang.ClassCastException: Expected instance of group converter but got "org.apache.spark.sql.execution.datasources.parquet.ParquetPrimitiveConverter"
     ```

  - on Trino:
     ```
    java.sql.SQLException: Query failed (#20230505_084023_00239_m8x8w): Error opening Hive split s3://trino-ci-test/databricks-compatibility-test-test_projection_pushdown_column_reorder_mz65kagqr1/part-00000-18f3b8e8-da96-4694-bb13-dcd4f1aa81df-c000.snappy.parquet (offset=0, length=4336): class org.apache.parquet.io.GroupColumnIO cannot be cast to class org.apache.parquet.io.PrimitiveColumnIO (org.apache.parquet.io.GroupColumnIO and org.apache.parquet.io.PrimitiveColumnIO are in unnamed module of loader io.trino.server.PluginClassLoader @1670ef43)
    ```

- With `name`:
  - on Delta: reads null values
  - on Trino: reads null values

#### When the data is inserted using `Trino`:
 - With `none`:
   - on Delta: Pass
   - on Trino: 
      ```
      io.trino.tempto.query.QueryExecutionException: java.sql.SQLException: Query failed (#20230505_090046_00274_m8x8w): Error opening Hive split s3://trino-ci-test/databricks-compatibility-test-test_projection_pushdown_column_reorder_9m0hicj4ka/20230505_090041_00271_m8x8w-146986f2-f931-48b2-a3d8-3b8745eeb49c (offset=0, length=935): Incorrect file size (935) for file (end of stream not reached): s3://trino-ci-test/databricks-compatibility-test-test_projection_pushdown_column_reorder_9m0hicj4ka/20230505_090041_00271_m8x8w-146986f2-f931-48b2-a3d8-3b8745eeb49c
     ```
 - With `id`: INSERT Not supported
 - With `name` column mapping mode,
   - on Delta: reads null values
   - on Trino:
     ```
      Caused by: java.sql.SQLException: Query failed (#20230505_090324_00281_m8x8w): Error opening Hive split s3://trino-ci-test/databricks-compatibility-test-test_projection_pushdown_column_reorder_opobubvb4d/20230505_090320_00278_m8x8w-82abef73-74a8-4302-a80d-6778f7052bb3 (offset=0, length=1743): Incorrect file size (1743) for file (end of stream not reached): s3://trino-ci-test/databricks-compatibility-test-test_projection_pushdown_column_reorder_opobubvb4d/20230505_090320_00278_m8x8w-82abef73-74a8-4302-a80d-6778f7052bb3
     ```

CC: @findinpath | @findepi

Reply:
It is normal that the test will fail when using column mappings.
The reason behind this is that the two different tables will have different column names when using column mapping `id` or `name`.

We probably need a different approach to test this scenario.
IMO The files need to be created manually - maybe `SHALLOW CLONE` Delta Spark feature would help in such a situation. (it would need though https://github.com/trinodb/trino/pull/17038)

@findepi  can we address this one as a follow-up?

Reply:
> i don't see yet how this helps. do you have a more complete code example?

I scratched together a test to:
- generate Parquet files via  trino's `ParquetWriter`
- upload the Parquet files to s3/MinIO
- use 
```
create table tiny.parquet_table_clone shallow clone parquet.`s3://trino-ci-test/parquet-files/`
```
However, the main problem with this method is that i can't specify the column mapping for the cloned table. 
So, yes, cloning is not necessarily an option.